### PR TITLE
Model Examples and LLM improvements

### PR DESCRIPTION
**Pull request by alfredorubin96**
Based on #591 

TODO:
- [x] improving schema fetch by using apoc.meta.data
- [x] changing prompt based on the latest examples
- [x] creating interface to provide examples to the model to improve accuracy

### CODE REVIEW

File: src/extensions/query-translator/clients/ModelClient.ts [hunk 1]
```diff
@@ -105,6 +120,33 @@ export abstract class ModelClient {
     return '';
   }
 
+  getExamplePrompt(examples) {
+    let res = `Here some samples of questions and their answers: \n`;
```
Comment (AleSim94) [inline src/extensions/query-translator/clients/ModelClient.ts:124]: rephrase this sentence

File: src/extensions/query-translator/clients/ModelClient.ts [hunk 2]
```diff
@@ -105,6 +120,33 @@ export abstract class ModelClient {
...
+
+  async manageMessageHistory(database, schema, schemaSampling, inputMessage, tmpHistory, reportType, examples) {
+    let res = [...tmpHistory];
+    // If empty, the first message will be the task definition
+    if (tmpHistory.length == 0) {
+      // The schema can be fetched in full or in sample mode (the second one is faster but less accurate)
+      schema = schemaSampling ? await this.generateSchemaSample(database) : await this.generateSchema(database);
+      tmpHistory.push(this.addSystemMessage(this.getTaskDefinition(schema)));
+    }
+    console.log(examples);
```
Comment (AleSim94) [inline src/extensions/query-translator/clients/ModelClient.ts:137]: I assume this is just for testing?

File: src/extensions/query-translator/clients/ModelClient.ts [hunk 3]
```diff
@@ -147,10 +196,10 @@ export abstract class ModelClient {
         retries += 1;
         onRetry(retries);
 
-        // Get the answer to the question
-        let modelAnswer = await this.chatCompletion(tmpHistory);
+        // Get the answer to the question from the model
+        modelAnswer = await this.chatCompletion(tmpHistory);
         tmpHistory.push(modelAnswer);
-
+        console.log(tmpHistory);
```
Comment (AleSim94) [inline src/extensions/query-translator/clients/ModelClient.ts:202]: probably also for testing

File: src/extensions/query-translator/clients/OpenAi/OpenAiClient.ts [hunk 4]
```diff
@@ -116,9 +116,9 @@ export class OpenAiClient extends ModelClient {
    */
   addUserMessage(content, reportType, plain = false) {
     let queryExample = reportExampleQueries[reportType];
-    let finalMessage = `${content}. Please use the following query structure as an example for ${reportTypesToDesc[reportType]}:
-  ${queryExample} 
-  Remember to respect the schema and remove any unnecessary comments or explanations from your result. Remember that every $ prefixed word is a parameter.`;
+    let finalMessage = `${content}. Please respect the structure of the result based on this description: ${reportTypesToDesc[reportType]}.
+  Here an example of query: ${queryExample}.
```
Comment (AleSim94) [inline src/extensions/query-translator/clients/OpenAi/OpenAiClient.ts:120]: "Here is an example of a query:"

File: src/extensions/query-translator/component/ClientSettings.tsx [hunk 5]
```diff
@@ -53,6 +56,10 @@ export const ClientSettings = ({
     entry[field] = value;
     setFunction(update(stateObj, entry));
   };
+  // TODO: REMOVE THIS JUST FOR TESTING
```
Comment (AleSim94) [inline src/extensions/query-translator/component/ClientSettings.tsx:59]: do not forget to remove

File: src/extensions/query-translator/clients/const.ts [hunk 6]
```diff
@@ -77,47 +110,18 @@ export const reportExampleQueries = {
...
-      // For each possible target
-      rel.target.forEach((trg) => res.push(`(:${rel.source})-[:${rel.relationship}]->(:${trg})`));
-    });
-    return res.join(',');
-  } catch (e) {
-    console.log(e);
-  }
-}
+// used for testing to remove
+export const MODEL_EXAMPLES_TEST = [
```
Comment (AleSim94) [inline src/extensions/query-translator/clients/const.ts:116]: do not forget to remove

**Review by AleSim94 — APPROVED** [review]

File: src/extensions/query-translator/clients/ModelClient.ts [hunk 7]
```diff
@@ -120,23 +160,30 @@ export abstract class ModelClient {
     history,
     database,
     reportType,
+    examples,
     onRetry = (value) => {
       let x = value;
-    }
+    },
+    schemaSampling = true // By default we create the schema message using apoc.meta.data in sampling mode
```
Comment (nielsdejong) [inline src/extensions/query-translator/clients/ModelClient.ts:167]: is this ever `false`?
Should this be a setting for the user inside the Modal?

**Review by nielsdejong — CHANGES_REQUESTED** [review]

File: gallery/tailwind.config.js [hunk 8]
```diff
@@ -6,7 +6,7 @@ module.exports = {
   theme: {
     extend: {},
   },
-  plugins: [],
+  plugins: [require('@tailwindcss/forms')],
```
Comment (nielsdejong) [inline gallery/tailwind.config.js:9]: why change from needle to this one?

File: src/extensions/query-translator/component/OverrideCardQueryEditor.tsx [hunk 9]
```diff
@@ -132,24 +141,33 @@ export const NeoOverrideCardQueryEditor = ({
...
-                      if (prepopulateExtensionName !== undefined) {
-                        triggerTranslation();
-                        setLanguage(Language.CYPHER);
-                        deletePrepopulationReportFunction(reportId);
-                      }
-                    }}
-                  >
-                    Translate
-                  </Button>
+                  <div style={{ display: 'flex', justifyContent: 'flex-end' }}>
```
Comment (nielsdejong) [inline src/extensions/query-translator/component/OverrideCardQueryEditor.tsx:148]: the UX might be confusing here for users, they could wonder
- What's a Q&A?
- Are they global or specific to this report?
- Should I bother creating them?

Perhaps this should all be in the extension config modal. 
How about a second screen opened by the extension modal called 'Tweak Model', where users can add examples etc.?

File: package.json [hunk 10]
```diff
@@ -88,6 +89,7 @@
     "react-markdown": "^8.0.0",
     "react-redux": "^7.2.6",
     "react-show-more-text": "^1.6.2",
+    "react-table": "^7.8.0",
```
Comment (BennuFire) [inline package.json:92]: Why another non needle dependency? We are trying to reduce imports

**Review by BennuFire — CHANGES_REQUESTED** [review]

File: package.json (hunk 10 above)
Comment (jacobbleakley-neo4j) [reply package.json:92]: As far as I am aware, react-table is necessary to implement pagination in the table which displays prompts/cypher queries. I haven't found a needle equivalent

Comment (BennuFire) [reply package.json:92]: Worst case scenario, there's @mui/x-data-grid. However, I do rather prefer using https://storybook-components-build.appspot.com/?path=/docs/tables-how-to-use--docs We will be soon removing x-data-grid.

We will not add any new style lib @alfredorubin96

File: package.json [hunk 11]
```diff
@@ -57,7 +57,8 @@
     "@nivo/sunburst": "^0.83.0",
```
Comment (nielsdejong) [inline package.json:57]: (Ignore the location of this comment)

We need to add some quick details in the documentation before merging this update.
Nothing too crazy, just a couple paragraphs with examples on the translator page.

(While we're at it, maybe we can rename the extension to `Text2Cypher` to match the language now used by the engineering team?)

File: package.json (hunk 10 above)
Comment (nielsdejong) [reply package.json:92]: Agree lets not add `@tanstack/react-table` and `react-table`

**Review by nielsdejong — APPROVED** [review]

**Pull request merged**